### ISSUE TITLE
protocols.py should conform to docstring and whitespace standards

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6654]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)| @staceysern reported|
|-|-|
|Trac ID|trac#6654|
|Type|enhancement|
|Created|2013-07-29 23:16:06Z|
|Branch|https://github.com/twisted/twisted/tree/protocol-doc-6654|

The docstrings in protocol.py should be updated to provide descriptions of all classes, functions and methods as well as their parameters and attributes. The spacing before classes and functions should conform to the coding standards.

Attachments:

* [protocol-doc-6654-22.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c3/2c35445b2a6b84cbc849a2d89f78aefff35c6c3e/1858958399940f7e14449c63783ecb29d0f59ba2.patch) (4040 bytes) - added by jefflovejapan on 2013-12-02 21:59:28Z - Updated tuple notation by 0-indexing and removing escape characters

<details><summary>Searchable metadata</summary>

```
trac-id__6654 6654
type__enhancement enhancement
reporter__shira shira
priority__normal normal
milestone__ 
branch__branches_protocol_doc_6654 branches/protocol-doc-6654
branch_author__shira shira
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__documentation documentation
time__1375139766000000 1375139766000000
changetime__1389501391000000 1389501391000000
version__None None
owner__shira shira

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_1"></a>|@staceysern commented|
|-|-|

(In [39376]) Branching to 'protocol-doc-6654'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_3"></a>|@staceysern commented|
|-|-|

(In [39377]) Fixed docstrings and whitespace.

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_4"></a>|@staceysern commented|
|-|-|

(In [39382]) Remove trailing whitespace.

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set owner to @staceysern|
|-|-|

Thanks Shira,

The new documentation looks good. I don't know twisted.mail very well
so this is just a cosmetic review.

Notes:
 * Build Results
   * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/protocol-doc-6654
   * There's a failure in debian-easy-no-zope-py2.6-epoll but I think
     its spurious.
     * https://buildbot.twistedmatrix.com/builders/debian-easy-no-zope-py2.6-epoll/builds/965/steps/epoll/logs/problems
 * Merges cleanly

Here are a few minor points:
 1. branches/protocol-doc-6654/twisted/mail/protocols.py
    1. Why did you remove the L{twisted.mail} link? I prefer it to
       just "mail" which is ambiguous.
    2. Use L{} for builtin types. eg C{str} should be L{str},
       C{NoneType} should be L{types.NoneType} or C{None}. L{dict}. etc
       1. Here's an example of good @type documentation
          https://twistedmatrix.com/documents/current/api/twisted.conch.endpoints.SSHCommandClientEndpoint.html#newConnection
    3. I think quite a few of the parameters and attibutes should be
       L{bytes} rathern than L{str}. eg protocolName, user, host,
       password, domainSpecifier, digest, filename
    4. "A domain delivery base class": Is that correct?
       L{SMTPDomainDelivery} seems to be a "super class" of
       L{DomainDeliveryBase}.
    5. Same for ESMTPDomainDelivery
    6. C{IMessageDelivery} > L{IMessageDelivery}
    7. "Positional arguments for L{SMTP}" should that be L{SMTP.__init__}?
       1. https://twistedmatrix.com/documents/current/api/twisted.mail.smtp.SMTP.html#__init__
    8. Same for ESMTP in DomainESMTP
    9. "173         @type protocol: L{SMTP} C{class} " No need to say C{class}.
       1. Same goes for POP3Factory
    10. "175            default value is L{SMTP} " Missing full stop.
    11. "218            default value is L{ESMTP} " Missing full stop.
    12. "224        @type challengers: C{dict} of C{str} -> C{class} which implements
        225             an L{ICredentials} sub-interface. " I found
        this confusing. Maybe just L{dict}. The following @ivar
        sentence describes the dictionary  adequately.
    13. "235            @param args: Arguments for L{SMTPFactory} "
        Link to L{SMTPFactory.__init__}.
    14. "247            @type addr: L{IAddress} provider": drop "provider"
    15. "290            @rtype: L{Deferred} which successfully results
        in": again, @rtype: L{Deferred} is enough. The following
        @ivar adequately describes the callback and errback circumstances.
    16. "351            @rtype: (C{str}, C{str}) " Should that be L{tuple}?
    17. "215                This loads a certificate and private key
        from a specified file. ": seems like a useful
        description. Why remove it?

 2. topfiles/6654.doc
    1. "api" > "API"
    2. I'd say "twisted.mail.protocols now has full API documentation."


Please resubmit for review after addressing the numbered points above
and remember to link to some new clean build results - especially the
documentation builder which should show any incorrect API links.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_7"></a>|@staceysern commented|
|-|-|

(In [39772]) Revised API documentation

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_8"></a>|@staceysern commented|
|-|-|

Fixed everything mentioned with the following exceptions or notes:

Issue 1: I removed the link to twisted.mail in the module docstring because that docstring appears on the twisted.mail package api documentation page which is what the link points to.  I also prefer seeing the docstrings use plain English except in the type descriptions.

Issue 4 & 5: SMTPDomainDelivery and ESMTPDomainDelivery inherit from DomainDeliveryBase so DomainDeliveryBase is the superclass.  SMTPDomain inherits from SMTPDomainDelivery so SMTPDomainDelivery is also a base class.  Similarly for ESMTPDomain and ESMTPDomainDelivery.  

Issue 9: L{SMTP} indicates an object of type L{SMTP} but the type here is the type of the class.  I changed it to use L{type} of instead of C{class}.

Issue 12: This is a little confusing because of how ICredentials are used but I cleaned it up a bit.

Line 247: There is a difference between an IAddress and an IAddress provider.  IAddress is a class which is an interface.  An IAddress provider is a class which implements the IAddress interface.  The latter is the type that is required here.  It's a little wordy but more accurate.  Most of the time the references are for the latter but occasionally, an interface is passed so it is important to distinguish the two.

Line 290: I know it's wordy but the detail on the types is important.  If we didn't have it and relied just on the text description, there would be no way to know that a logout function is a no argument callable or that a mailbox is of type pop3.Mailbox.

Line 351: In response to this comment, I came up with a much better way of documenting tuples and fixed length lists that removes a lot of ambiguity.

Line 215: The description seemed more appropriate for the getContext method.  

Also added missing docstrings for the following DomainDeliveryBase functions: receivedHeader, validateTo, validateFrom. Added docstrings for DomainSMTP.__init__ and DomainESMTP.__init__.

Additional changes were made to incorporate feedback from other documentation reviews and for general improvement in readability.  They include:
 * Changing type references from str to bytes
 * Including a locally usable name in the link
 * Removing blank lines after class docstrings
 * Removing (optional) in param descriptions.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_11"></a>|@staceysern commented|
|-|-|

(In [39805]) Revised API documentation

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_12"></a>|@staceysern commented|
|-|-|

Found a better way to deal with issue 9.  It's a no-argument callable which returns L{SMTP}.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Replying to [shira](#note_8):
> Fixed everything mentioned with the following exceptions or notes:
> 
> Issue 1: I removed the link to twisted.mail in the module docstring because that docstring appears on the twisted.mail package api documentation page which is what the link points to.  I also prefer seeing the docstrings use plain English except in the type descriptions.

We use this 'convention' in the tests as well so I'd also rather keep the L{twisted.mail} link.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_14"></a>|@staceysern commented|
|-|-|

(In [39949]) Docstring revisions

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_15"></a>|@staceysern commented|
|-|-|

On issue 1, I changed the docstring again to fit in better with the rest of the mail module level docstrings and avoid the need for a link.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_16"></a>|@staceysern commented|
|-|-|

(In [40102]) Fix incorrect docstring

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_17"></a>|@staceysern commented|
|-|-|

(In [40116]) Remove extraneous letter

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_18"></a>|@lvh set owner to @lvh|
|-|-|

Hi shira! Thanks for your extensive work on twisted.mail, both on this ticket and elsewehere :)

Not directly related to your patch, but why does ```DomainDeliveryBase``` set ```service = None``` in class scope when ```__init__``` sets it unconditionally? This may be an opportunity to kill it :)

I thought there weren't supposed to be empty newlines between epytext var/attr markup?

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_19"></a>|@lvh commented|
|-|-|

Sorry, I didn't mean to send that yet.

Not a review requirement since it's not an actual rule, but is putting @type before the @param intentional? It seems backwards to me.

There are some remaining twistedchecker failures:

```
************* Module twisted.mail.protocols
C0103: 84,8:DomainDeliveryBase.receivedHeader: Invalid name "from_" (should match ([a-z_][a-zA-Z0-9]*)$)
C0301: 88,0: Line too long (83/79)
C0103: 88,8:DomainDeliveryBase.receivedHeader: Invalid name "for_" (should match ([a-z_][a-zA-Z0-9]*)$)
C0301:136,0: Line too long (82/79)
C0301:138,0: Line too long (88/79)
C0301:208,0: Line too long (80/79)
W0311:422,0: Bad indentation. Found 13 spaces, expected 12
W9501:422,34:VirtualPOP3.lookupDomain: String formatting operations should always use a tuplefor non-mapping values
```

While the invalid name ones are arguably twistedchecker bugs, it may be a good idea to name them something more descriptive anyway.

Thanks again for working on this!

cheers
lvh

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern set owner to @staceysern|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="jefflovejapan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|jefflovejapan commented|
|-|-|

Updated tuple notation by 0-indexing and removing escape characters.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [41110]) Apply protocol-doc-6654-22.patch 

refs [#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This is a great documentation improvement.  I somewhat agree with the idea that links like `L{twisted.mail}` are useful but I don't think the change to remove them here is a big deal.

I do still find the tuple type documentation confusing but apart from that I think this change is ready to land.  I'll make the slight adjustment to that part of the docs and then merge the branch.  Thanks all!

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_25"></a>|@staceysern commented|
|-|-|

(In [41112]) Modify docstring and remove extra blank lines

refs=[#6654](https://github.com/twisted/twisted/issues/6654)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @staceysern|
|-|-|

shira suggested she'd finish this up.

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_27"></a>|@staceysern set status to `closed`|
|-|-|

(In [41208]) Merge protocol-doc-6654: Fix docstrings for protocols.py

Author: shira
Reviewer: rwall, exarkun
Fixes: [#6654](https://github.com/twisted/twisted/issues/6654)

Long description (as long as you wish)
